### DISCUSSION TITLE
Allow CLI `--output=output.html` to override permalink of `--input=input.html`

### DISCUSSION THREAD

**Discussion by thecristen** [11ty/eleventy#3995] [Enhancement Queue]
**Describe the bug**
I'm using Eleventy on the command line to compile a Nunjucks file into an HTML file. It works great but I'm having trouble getting the output to go where I want to. I want to compile the `test.njk` file I have in the root directory to `test.html` in the same directory. I've read the using the same [input and output directory pitfall page](https://www.11ty.dev/docs/languages/html/#same-input-output), but I don't think it applies here.

**To Reproduce**
Contents of `test.njk` below, all of the included files are simple HTML files:

```njk
<!-- 
  I've got some redacted comments here
 -->
<html lang="en">
  <head>
  {% include "head.html" %}
  <body>
    <div class="body-wrapper" id="body-wrapper">
      {% include "header.html" %}
      <hr>
      {% include "footer.html" %}
    </div>
  {% include "scripts.html" %}
  </body>
</html>
```

Trying to compile:

```
eleventy --dryrun --input=test.njk --output=. 
[11ty] Skipping ./test/index.html from ./test.njk

eleventy --dryrun --input=test.njk --output=test.html
[11ty] Skipping test.html/test/index.html from ./test.njk

eleventy --dryrun --input=test.njk --output=./test.html
[11ty] Skipping ./test.html/test/index.html from ./test.njk
```

**Expected behavior**

I want `test.html` output to the root directory, not under a `test` or `test.html` folder.
Interestingly, it worked back when I had the file named `index.html`.

```
eleventy --input=index.njk --output=.          
[11ty] Writing ./index.html from ./index.njk
[11ty] Wrote 1 file in 0.10 seconds (v1.0.0)
```

**Environment:**

- OS and Version: macOS Monterey 12.1
- Eleventy Version 1.0.0

**Comment by pdehaan** [comment]
Actually, rereading your original comment of _"I want to compile the test.njk file I have in the root directory to test.html in the same directory."_, I think that pitfall _does_ potentially apply (assuming Eleventy tries to parse the generated HTML files on subsequent builds). Or if you're strictly using Eleventy to build single templates and output to the same input directory (in which case the caveat might NOT apply).

My frontmatter comment _might_ still stand. I guess it depends on how you're building.
Personally I find using the same input+output directories confusing, and I prefer to put all my templates in a subdirectory and then set the `dir.input` and `dir.output` directories in my .eleventy.js config file; but I understand that might not work for everybody's use cases.

If you want to throw your project up in a public GitHub repo I can take a look and see if we can get it building how you want it. I'm a bit curious about the HTML partials above for "head.html", "header.html", "footer.html", and "scripts.html".

~~@thecristen Nope, I don't think that pitfall applies to this, but you'll probably want to check out the https://www.11ty.dev/docs/permalinks/ docs.~~

~~TLDR: you can add a `permalink: test.html` to your frontmatter to change the output file on a per-template basis. But that can be tedious for large sites.
I'd have to play around with it a bit, but for multiple pages, we could set it in a directory data file, or possibly something like a global `_data/eleventyComputed.js` data file.
But yes, the index.xxx pages are a bit of a special case, as you see above.~~

**Comment by thecristen** [comment]
Gotcha. I'm processing no other files in the entire project, so I was hoping to not have to add front matter or any other 11ty-specific configuration to the project to process the one file. 

A use case to consider, perhaps, but good to know that the behavior I ran into is expected 🥲

**Comment by pdehaan** [comment]
Ah, updated my comment above.

Also interesting point from your output above:

```sh
eleventy --dryrun --input=test.njk --output=test.html
[11ty] Skipping test.html/test/index.html from ./test.njk
```

So another gotcha is the [<kbd>--input</kbd> flag](https://www.11ty.dev/docs/config/#input-directory)

> _"Controls the top level **directory/file/glob** that we’ll use to look for templates."_

Whereas the [<kbd>--output</kbd> flag](https://www.11ty.dev/docs/config/#output-directory)

> _"Controls the **directory** inside which the finished templates will be written to."_

Which you can see from your output above: "test.html/test/index.html"

I'm still struggling to come up with a workaround that isn't `permalink: test.html` frontmatter in the test.njk template, or a directory data file like "test.11tydata.json" which applies frontmatter for that one specific file (but is one more file in your project). Or else some post-build script which simply renames/moves the generated ./test/index.html file to ./test.html, which would probably be easy if you already have a package.json file in the project.

**Comment by thecristen** [comment]
Ahh good note about the documented difference in how the input and output flags work. Thank you for that.

> Whereas the `--output` flag
> > _"Controls the **directory** inside which the finished templates will be written to."_

Noted! But doesn't this suggest that `--output=.` _should_ output into the top level directory, then? 🤔

**Comment by pdehaan** [comment]
> But doesn't this suggest that `--output=.` _should_ output into the top level directory, then? 🤔

Ah ha! Except it is writing to the top level directory! It's writing the output to "test/index.html" in the top-level directory (which is the same level as your .eleventy.js config file, or the ./test.njk file).

I have this working locally:

```njk
---
permalink: test.html
---

<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
</head>
<body>

{% include "content.html" %}

</body>
</html>
```

Where my package.json "build" script looks like <kbd>eleventy --input=test.njk --output=.</kbd>.
But if you don't want to add front matter (or can't for whatever reason), I think you might be stuck with the file being written to ./test/index.html and then moving/renaming it to the desired name/location.

For example, here's my package.json `scripts` section (when not using the `permalink` solution above):

```js
  "scripts": {
    "build": "eleventy --input=test.njk --output=.",
    "postbuild": "mv test/index.html test.html",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
```

Now when I run <kbd>npm run build</kbd> it will run the `build` script (which builds the one Nunjucks template), and then the `postbuild` hook will automatically run after the <kbd>npm run build</kbd> script and move the file to the desired place.

I guess a final option might be seeing if you can use `nunjucks-cli` or something similar. I'm not sure which features of Eleventy you're using, but given my super simple example of just using `{% include "xxx" %}` in the simple templates above, I was able to compile with vanilla Nunjucks using:

```sh
npx nunjucks-cli "*.njk" -p ""
```

Not sure why I had to use that weird `-p ""` hack, but without it, I was getting some errors:

```sh
npx nunjucks-cli "*.njk"
node:internal/validators:120
    throw new ERR_INVALID_ARG_TYPE(name, 'string', value);
    ^

TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
    at new NodeError (node:internal/errors:371:5)
    at validateString (node:internal/validators:120:11)
    at resolve (node:path:1098:7)
    at Object.<anonymous> (/private/tmp/11ty-2184/node_modules/nunjucks-cli/main.js:55:18)
    at Module._compile (node:internal/modules/cjs/loader:1097:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1149:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47 {
  code: 'ERR_INVALID_ARG_TYPE'
}

Node.js v17.3.1
```

I'll file a bug in https://github.com/jeremyben/nunjucks-cli, but not sure how active that repo is. :shrug:
**UPDATE:** Filed as https://github.com/jeremyben/nunjucks-cli/issues/26

**Comment by zachleat** [comment]
Ah, this is the first time I’ve seen a request to override the permalink of `--input=file` with `--output=file`. I think it has some merit but unfortunately I don’t think I’ve seen other folks face this confusion. I’ll put it in the enhancement queue, thanks!

> This repository is now using [lodash style issue management](https://twitter.com/samselikoff/status/991395669016436736) for enhancements. This means enhancement issues will now be closed instead of leaving them open.

> View [the enhancement backlog here](https://github.com/11ty/eleventy/issues?utf8=%E2%9C%93&q=label%3Aneeds-votes+sort%3Areactions-%2B1-desc+). Don’t forget to upvote the top comment with 👍!
